### PR TITLE
Poison address-exposed user variables in debug

### LINKED ISSUE #13072: Fill uninitialized address-taken locals with poison pattern in debug codegen
C/C++ compilers have this debugging/diagnostic aid for years.

More context: https://github.com/dotnet/coreclr/pull/25674#issuecomment-511118490

Related to: https://github.com/dotnet/csharplang/issues/868

category:testing
theme:testing
skill-level:beginner
cost:small

### PR DESCRIPTION
**Pull request by jakobbotsch**
Fix #13072

### CODE REVIEW

**Comment by jakobbotsch** [comment]
@sandreenko I hope I have addressed what we talked about some weeks ago.
1. Now we only poison address exposed user locals
2. The poison code is emitted in a scratch BB right after the prolog
3. It's implemented for ARM/ARM64 as well

PTAL

**Comment by sandreenko** [comment]
cc @dotnet/jit-contrib

**Comment by jakobbotsch** [comment]
I want to run local tests first before I do some more testing, not sure that this is quite ready yet.

File: src/coreclr/jit/codegencommon.cpp
Comment (sandreenko) [inline src/coreclr/jit/codegencommon.cpp:12557]: I would like to have them defined in the target*.h files, like `#define REG_FRAME_POISONING`.

File: src/coreclr/jit/codegencommon.cpp
Comment (sandreenko) [inline src/coreclr/jit/codegencommon.cpp:12603]: Nit: I would prefer to avoid  unnamed constants, maybe similar to what we do in `genCodeForCpBlkUnroll` https://github.com/dotnet/runtime/blob/1d84e014ac2b6d11b1ddc452ce1a7706395abcb6/src/coreclr/jit/codegenarmarch.cpp#L2206 
```
const unsigned longRegSize = genTypeSize(TYP_LONG);
if ((offs % longRegSize  == 0) && (end - offs >= longRegSize))
...
offs += longRegSize;
...
```

**Review by sandreenko — COMMENTED** [review]
LGTM, this is an elegant solution, a few nits/questions.

File: src/coreclr/jit/codegencommon.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp:12557]: This has to be changed anyway since we can home arguments into registers in the prolog, depending on RA.

File: src/coreclr/jit/codegencommon.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp:12602]: I think it is still suggested to use aligned writes since there may be penalties for unaligned writes that cross cache lines, no?

File: src/coreclr/jit/compiler.h
Comment (jakobbotsch) [reply src/coreclr/jit/compiler.h:9828]: Can you explain exactly what `compDbgCode` entails? Does it mean user has attached a debugger? What if user attaches debugger after method was compiled in debug?

File: src/coreclr/jit/flowgraph.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/flowgraph.cpp:2608]: The pinvoke frame prolog is generated as IR, so the poisoning is emitted first.

File: src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs
Comment (jakobbotsch) [reply src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs:156]: You mean instead of using 0xcc? Sure, I can change that. But the libraries seem a bit inconsistent. Some places uses 0xcc for this kind of debug filling and MSVC does it too, so that's why I went with it.

File: src/coreclr/jit/codegencommon.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp:12603]: IMO, these constants are so fundamental that there is not much to be gained doing that. `EA_8BYTE` is just as unnamed as `8` in my opinion. But I can change it if you still think it's better.

File: src/coreclr/jit/compiler.h
Comment (sandreenko) [reply src/coreclr/jit/compiler.h:9828]: It means the C#/IL code was compiler with "Debug" defined, so in VS it is controlled here:
image

File: src/coreclr/jit/codegencommon.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp:12557]: Turns out there is already `REG_SCRATCH` defined that I will reuse for this.

File: src/coreclr/jit/compiler.h
Comment (jakobbotsch) [reply src/coreclr/jit/compiler.h:9828]: In that case I don't think it should be necessary, I assume `compDbgCode` will imply that `MinOpts()` is true anyway?

Comment (sandreenko) [reply src/coreclr/jit/compiler.h:9828]: I would say they are independent, so you can have any combinations of "compDbgCode" and "MinOpts()". 
There is such method:
```
        bool OptimizationDisabled() const
        {
            return MinOpts() || compDbgCode;
        }
```
but I do not think you want to use it here, I would just leave `return !info.compInitMem && opts.compDbgCode;`

**Review by nelsondani — APPROVED** [review]

File: src/coreclr/jit/compiler.h
Comment (jakobbotsch) [reply src/coreclr/jit/compiler.h]: Got it, will do that.

**Comment by jakobbotsch** [comment]
This should be ready for review. I believe the failing tests are #54778.
To make sure we do not allocate things to the scratch register being used for the immediate I have added a kill in register allocation when we see the scratch BB and know that we will emit poisoning. @kunalspathak can you take a look at this part?

cc @dotnet/jit-contrib

File: src/coreclr/jit/codegencommon.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp:12578]: Intel's optimization reference says in 3.6.6 that there is a penalty when unaligned stack access splits a cache line, so I will leave this as is.

File: src/coreclr/jit/codegencommon.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp]: I switched to modulo but kept `4` and `8`.

File: src/coreclr/jit/flowgraph.cpp
Comment (kunalspathak) [inline src/coreclr/jit/flowgraph.cpp:2606]:
```suggestion
    // The backend requires a scratch BB into which it can safely insert a P/Invoke method prolog if one is
    // required. Similarly, we need a scratch BB for poisoning. Create it here.
```

**Review by kunalspathak — COMMENTED** [review]
Could you also include a test for this?

**Review by kunalspathak — APPROVED** [review]
LGTM

File: src/coreclr/jit/flowgraph.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/flowgraph.cpp:2607]:
```suggestion
```

**Comment by jakobbotsch** [comment]
I added a test. Failing tests are #54778 and some Mono failures (looks like https://github.com/dotnet/runtime/issues/53329#issuecomment-871039590). I'll wait for @echesakov's review before merging.

**Review by sandreenko — APPROVED** [review]
LGTM

**Review by echesakov — APPROVED** [review]
LGTM

**Pull request merged**
